File: 15-3sum/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res;
        sort(nums.begin(),nums.end());
        for(int i=0;i<nums.size()-1;i++){
            if(i!=0 && nums[i]==nums[i-1]){
                continue;
            }
            int start = i+1;
            int end = nums.size()-1;
            while (start < end){
                int curr = nums[start] + nums[end] + nums[i];
                if(curr>0){
                    end--;
                }
                else if(curr<0){
                    start++;
                }
                else{
                    res.push_back({nums[i],nums[start],nums[end]});
                    while(start < end && nums[end]==nums[end-1]){
                        end--;
                    }
                    while(start < end && nums[start]==nums[start+1]){
                        start++;
                    }
                    start++;
                    end--;
                }
            }
        }
        return res;
    }
};

File: 15-3sum/3sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """

        res = []

        nums.sort()
        for i in range (len(nums)):
            
            l, r = i + 1, len(nums) - 1
            if (i > 0  and nums[i] == nums[i-1]):
                continue 
            while l < r:
                
                if (nums[i] + nums[l] + nums[r] == 0):
                    temp = [nums[i], nums[l], nums[r]]
                    res.append(temp)
                    
                    while (l < r and nums[l] == nums[l+1]):
                        l += 1
                    
                    while (l < r and nums[r] == nums[r-1]):
                        r -= 1

                    l += 1
                    r -= 1
                elif (nums[i] + nums[l] + nums[r] > 0):
                    r -= 1
                
                else:
                    l += 1
        
        return res
        